mbed_devices/_internal/windows/component_descriptor.py

```diff
@@ -0,0 +1,123 @@
+"""Defines a generic Win32 component."""
+import logging
+from abc import ABC, abstractmethod
+from typing import List, Any, Generator, Optional
+
+import win32com.client
+
+UNKNOWN_VALUE = "Unknown"
+NAMED_TUPLE_FIELDS_ATTRIBUTE = "_fields"
+
+logger = logging.getLogger(__name__)
+
+
+class ComponentDescriptor(ABC):
+    """Win32 component descriptor."""
+
+    @property
+    @abstractmethod
+    def class_name(self) -> str:
+        """Returns the name of the Win32 Class."""
+        return UNKNOWN_VALUE
+
+    @property
+    def parameters(self) -> List[str]:
+        """Returns all the different parameters of the class."""
+        return self._class_parameters()
+
+    @classmethod
+    def _class_parameters(cls) -> List[str]:
+        return [k for k in getattr(cls, NAMED_TUPLE_FIELDS_ATTRIBUTE)]
+
+    @property
+    @abstractmethod
+    def device_id(self) -> str:
+        """Returns the device ID field."""
+        pass
+
+    @property
+    def filter(self) -> Optional[str]:
+        """Filter applied on the list of elements."""
+        return None
+
+    @property
+    def is_undefined(self) -> bool:
+        """Determines whether the structure is undefined or not i.e. none of the fields are actually defined."""
+        undefined_values = []
+        for param in self.parameters:
+            value = self.__getattribute__(param)
+            undefined_values.append(ComponentDescriptor._is_undefined_value(value))
+        return all(undefined_values)
+
+    @staticmethod
+    def _is_undefined_value(value: Any) -> bool:
+        return (
+            value.is_undefined if isinstance(value, ComponentDescriptor) else value in [None, UNKNOWN_VALUE, 0, False]
+        )
+
+    @staticmethod
+    def _read_cdispatch_field(win32_element: Any, key: str) -> Any:
+        try:
+            return getattr(win32_element, key)
```

Comment:
Why not just do `win32_element.key` if we're catching the `AttributeError` anyway?

Reply:
Oh, "key" is a string.